sdk/core/azure_core_amqp/src/cbs.rs

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) Microsoft Corp. All Rights Reserved.
+// cspell: words amqp sasl
+
+use azure_core::error::Result;
+use std::fmt::Debug;
+
+use super::session::AmqpSession;
+
+#[cfg(all(feature = "enable-fe2o3-amqp", not(target_arch = "wasm32")))]
+type CbsImplementation = super::fe2o3::cbs::Fe2o3ClaimsBasedSecurity;
+
+#[cfg(any(not(any(feature = "enable-fe2o3-amqp")), target_arch = "wasm32"))]
+type CbsImplementation = super::noop::NoopAmqpClaimsBasedSecurity;
+
+pub trait AmqpClaimsBasedSecurityTrait {
+    /// Asynchronously attaches the Claims-Based Security (CBS) node to the AMQP session.
+    ///
+    /// This method is responsible for setting up the necessary AMQP links for CBS operations.
+    /// It must be called before attempting to authorize any AMQP paths using the `authorize_path` method.
+    ///
+    /// # Returns
+    ///
+    /// A `Result` which is:
+    /// - `Ok(())` on successful attachment of the CBS node.
+    /// - `Err(e)` where `e` is an error from the `azure_core::error::Result` indicating the failure reason.
+    ///
+    fn attach(&self) -> impl std::future::Future<Output = Result<()>> {
+        async { unimplemented!() }
+    }
+
+    /// Asynchronously authorizes an AMQP path using the provided secret.
+    ///
+    /// The authorization is valid until the specified `expires_on` time. The path is typically a URI that represents an AMQP resource. The secret is typically a SAS token. The `expires_on` time is the time at which the authorization expires.
+    ///
+    /// # Parameters
+    ///
+    /// - `path`: A `String` reference representing the AMQP path to be authorized.
+    /// - `secret`: An implementor of `Into<String>` representing the secret used for authorization. This is typically a JSON Web token.
+    /// - `expires_on`: A `time::OffsetDateTime` representing the expiration time of the authorization.
+    ///
+    /// # Returns
+    ///
+    /// A `Result` which is:
+    /// - `Ok(())` on successful authorization of the AMQP path.
+    /// - `Err(e)` where `e` is an error from the `azure_core::error::Result` indicating the failure reason.
+    ///
+    #[allow(unused_variables)]
+    fn authorize_path(
+        &self,
+        path: impl Into<String> + Debug,
+        secret: impl Into<String>,
+        expires_on: time::OffsetDateTime,
+    ) -> impl std::future::Future<Output = Result<()>> {
+        async { unimplemented!() }
+    }
+}
+
+#[derive(Debug)]
+struct AmqpClaimsBasedSecurityImpl<T>(T);
+
+impl<T> AmqpClaimsBasedSecurityImpl<T>
+where
+    T: AmqpClaimsBasedSecurityTrait,
+{
+    pub(crate) fn new(cbs: T) -> Self {
```

Comment:
Tip: unless you end up exposing a type - or are worried it might be - outside of `src/lib.rs`, `pub(crate)` is overkill. Ultimately, you could make everything `pub` but if it's not `pub mod` or `pub use` in `src/lib.rs` it won't matter. This doesn't hurt, hence the "tip" prefix.

Reply:
Ah - declare the "new" function as "pub" because the containing struct is pub(crate) and thus pub(crate) is irrelevant on the new function.

I was worried about a "function `new` has greater visibility than containing struct" error, but that apparently doesn't happen.